http: Add support for appending request/response headers only if they are absent

agrawroh opened the pull request:
### Background

Currently, there seems to be no way to only add a request/response header if it doesn't already exist. Envoy will either append the value if the header already exists or will overwrite it depending on the value of `append` [flag](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#envoy-v3-api-msg-config-core-v3-headervalueoption) passed to the `request_headers_to_add` or `response_headers_to_add` [[Link]](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto).

We have a use-case where we only want to add a certain request/response header if it doesn't already exist and it could be very useful to also have a flag (say, `overwrite`) to control the behavior of overwriting the header value.

### Changes

This PR adds a new field called `append_action` and marks the old `append` to be deprecated in favor of this new field. `append_action` extends the header matcher by describing what append action to take indicating whether:
1. The formatted header should be added only if it's not already present;
2. New value should be appended to the existing values if header already exists; or
3. New value should be overwritten by discarding any existing values if header already exists;

**Commit Message:** added support for appending request/response headers only if they are absent.
**Additional Description:** Introduces a new flag called `overwrite` which can be used to control whether an existing header value should be overwritten or not.
**Risk Level:** Low
**Testing:** Unit Tests, Integration Tests
**Docs Changes:** Added the description of the new field which is added to support this feature. 
**Release Notes:** Added
**Platform Specific Features:** N/A

**Signed-off-by:** Rohit Agrawal <rohit.agrawal@databricks.com>

agrawroh commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -323,9 +323,17 @@ message HeaderValueOption {
+  // when `append` is false.
+  // If both the `append` and `overwrite` are set to false then the given header is only appended
+  // if it doesn't already exist.
+  google.protobuf.BoolValue overwrite = 3;
```
**Note to reviewers:**

I couldn't think of any other name here. If there are any other suggestions please let me know and I'll change it accordingly.

rojkov replied:
What would be the semantical meaning if both `append` and `override` were set to true?

Perhaps a better alternative would be an enum type instead of the set of bools. Something like `overrideExisting` (default), `appendToExisting` and `addIfMissing`.

soulxu replied:
or just rename this bool field to `add_if_missing`, then it only controls whether to add new header if there is no one. The enum type is good when we can remove the `append` field, since it looks like duplicated with enum type new field. But we can't remove the `append` for back-forward compatibility I guess.

lambdai replied:
From high level the options are 3 
replace-or-override
append
append-if-exist

I would suggest create new enum field and deprecate the existing `append`

mattklein123 replied:
+1 to deprecate and add an enum. I think this is confusing. I would just translate the existing append field into the enum and then offer the enum value for the new behavior. Thank you!

/wait

agrawroh replied:
Changed it to the `enum` as per the suggestion. Thanks!

wbpcode reviewed (commented):
Thanks working on this. This function also makes sense for me. Some comment to API.

lambdai commented on source/common/router/header_formatter.h:
```diff
@@ -35,18 +41,20 @@ using HeaderFormatterPtr = std::unique_ptr<HeaderFormatter>;
 class StreamInfoHeaderFormatter : public HeaderFormatter {
 public:
-  StreamInfoHeaderFormatter(absl::string_view field_name, bool append);
+  StreamInfoHeaderFormatter(absl::string_view field_name, bool append, bool overwrite);
```
Regardless the protobuf API is, we should use the enum here. 2 bools for 3 options are confusing

agrawroh replied:
Changed it to `enum` as per the suggestion. Thanks!

lambdai reviewed (commented):
My 2 cents: We should split api and c++ code

agrawroh commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -323,9 +323,17 @@ message HeaderValueOption {
+  // when `append` is false.
+  // If both the `append` and `overwrite` are set to false then the given header is only appended
+  // if it doesn't already exist.
+  google.protobuf.BoolValue overwrite = 3;
```
Changed it to the `enum` as per the suggestion. Thanks!

mattklein123 commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -320,12 +320,35 @@ message HeaderValueOption {
+  // Describes the supported actions types to be used in :ref:`append_action <envoy_v3_api_field_config.core.v3.HeaderValueOption.append_action>`
+  enum HeaderAppendAction {
+    // This action will append the specified value to the existing values if the header
+    // already exist. If the header doesn't exist then this will add the header with
```
```suggestion
    // already exists. If the header doesn't exist then this will add the header with
```

mattklein123 commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -320,12 +320,35 @@ message HeaderValueOption {
+    APPEND_IF_EXISTS = 0;
+
+    // This action will only append the header if it doesn't already exist. If the header
+    // already exist then this will be a no-op.
```
```suggestion
    // already exists then this will be a no-op.
```

mattklein123 commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -320,12 +320,35 @@ message HeaderValueOption {
+
+    // This action will only append the header if it doesn't already exist. If the header
+    // already exist then this will be a no-op.
+    APPEND = 1;
```
For clarity I would call this `ADD_IF_ABSENT` for clarity. That is what it is, right?

mattklein123 commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -320,12 +320,35 @@ message HeaderValueOption {
+    APPEND = 1;
+
+    // This action will overwrite the specified value by discarding any existing values if
+    // the header already exist. If the header doesn't exist then this will add the header
```
```suggestion
    // the header already exists. If the header doesn't exist then this will add the header
```

mattklein123 reviewed (changes requested):
Thanks this looks good from an API perpsective modulo a couple of small comments.

/wait

agrawroh commented:
> Thanks this looks good from an API perpsective modulo a couple of small comments.
> 
> /wait

Thanks for taking a look @mattklein123. I addressed the typos and changed the field to `ADD_IF_ABSENT` which I think is more clear.

mattklein123 commented on docs/root/version_history/current.rst:
```diff
@@ -134,6 +135,8 @@ Deprecated
   :ref:`prefix_match <envoy_v3_api_field_config.route.v3.HeaderMatcher.prefix_match>`, :ref:`suffix_match <envoy_v3_api_field_config.route.v3.HeaderMatcher.suffix_match>` and
   :ref:`contains_match <envoy_v3_api_field_config.route.v3.HeaderMatcher.contains_match>` are deprecated by :ref:`string_match <envoy_v3_api_field_config.route.v3.HeaderMatcher.string_match>`.
+* http: :ref:`append <envoy_v3_api_field_config.core.v3.HeaderValueOption.append>` field has been deprecated in favor of the new
+  :ref:`append_action <envoy_v3_api_field_config.core.v3.HeaderValueOption.append_action>` field which is more clear and can be used to further extend the header matcher functionality.
```
nit: can you see what functionality you actually added and ref link to that specifically?

mattklein123 commented on envoy/http/header_map.h:
```diff
@@ -686,8 +686,15 @@ using HeaderMapPtr = std::unique_ptr<HeaderMap>;
 struct HeaderTransforms {
-  std::vector<std::pair<Http::LowerCaseString, std::string>> headers_to_append;
-  std::vector<std::pair<Http::LowerCaseString, std::string>> headers_to_overwrite;
+  // List of headers which needs to be added if they are absent.
```
```suggestion
  // List of headers which need to be added if they are absent.
```

mattklein123 commented on envoy/http/header_map.h:
```diff
@@ -686,8 +686,15 @@ using HeaderMapPtr = std::unique_ptr<HeaderMap>;
+  // List of headers which needs to be added if they are absent.
+  std::vector<std::pair<Http::LowerCaseString, std::string>> headers_to_add_if_absent;
+  // List of headers for which the new value should be appended to the existing values if these
+  // headers already exist.
```
```suggestion
  // headers already exist. If the headers don't exist they are added.
```

mattklein123 commented on source/common/grpc/google_async_client_impl.cc:
```diff
@@ -83,8 +83,9 @@ GoogleAsyncClientImpl::GoogleAsyncClientImpl(Event::Dispatcher& dispatcher,
-          Router::HeaderParser::configure(config.initial_metadata(), /*append=*/false)) {
+      metadata_parser_(Router::HeaderParser::configure(
+          config.initial_metadata(),
+          /*append_action=*/envoy::config::core::v3::HeaderValueOption::OVERWRITE_IF_EXISTS)) {
```
You can remove `/*append_action=*/` as it's now obvious.

agrawroh replied:
Make sense. I removed it.

mattklein123 commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -320,12 +320,35 @@ message HeaderValueOption {
   // Should the value be appended? If true (default), the value is appended to
   // existing values. Otherwise it replaces any existing values.
-  google.protobuf.BoolValue append = 2;
+  // This field is now deprecated in the favor of :ref:`append_action <envoy_v3_api_field_config.core.v3.HeaderValueOption.append_action>`.
```
Can you document specifically which new enum values this true and false here map to?

agrawroh replied:
Done, Thanks!

mattklein123 commented on source/common/router/header_formatter.h:
```diff
@@ -22,10 +22,13 @@ class HeaderFormatter {
-   *              headers or replace any existing values for the header
+   * @return envoy::config::core::v3::HeaderValueOption::HeaderAppendAction describes what append
+   * action to take indicating whether:
+   * 1. The formatted header should be added only if it's not already present;
+   * 2. New value should be appended to the existing values if header already exists; or
+   * 3. New value should be overwritten by discarding any existing values if header already exists;
```
I don't think you need to re-document this. The enum is self explanatory.

agrawroh replied:
Removed.

mattklein123 commented on source/common/router/header_parser.cc:
```diff
@@ -227,8 +228,18 @@ HeaderParserPtr HeaderParser::configure(
+    envoy::config::core::v3::HeaderValueOption::HeaderAppendAction append_action =
+        header_value_option.append_action();
+    // Preserve the old behavior until `append` field is fully deprecated.
+    if (header_value_option.has_append()) {
```
If someone sets this but also has set append_action to a non-default value, I would raise an error (please test this).

agrawroh replied:
Done. Thanks!

mattklein123 commented on source/common/router/header_parser.cc:
```diff
@@ -286,12 +297,30 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
+    case envoy::config::core::v3::HeaderValueOption::ADD_IF_ABSENT:
+      // Check whether the header already exist or not. We only
+      // need to add the header if it doesn't already exist.
+      if (headers.get(key).empty() && !value.empty()) {
```
You check value.empty() at insertion time. Can you just assert here and below that it's not empty?

agrawroh replied:
Changed, Thanks!

agrawroh replied:
**UPDATE:** I realized that there are a few test cases which fails if I change this check to an assertion instead.

https://github.com/envoyproxy/envoy/blob/05ca792d25e4e40e472ba34ba4206fb69214929a/test/common/router/header_formatter_test.cc#L803

mattklein123 replied:
Can you debug why they are failing? Is it because of a test issue? Or is it an issue that can actually happen?

/wait

agrawroh replied:
@mattklein123 
Looks like these three test cases actually assert on this behavior:
https://github.com/envoyproxy/envoy/blob/05ca792d25e4e40e472ba34ba4206fb69214929a/test/common/router/header_formatter_test.cc#L1089
https://github.com/envoyproxy/envoy/blob/05ca792d25e4e40e472ba34ba4206fb69214929a/test/common/router/header_formatter_test.cc#L1081
https://github.com/envoyproxy/envoy/blob/05ca792d25e4e40e472ba34ba4206fb69214929a/test/common/router/header_formatter_test.cc#L812

It doesn't make much sense to supply an empty value here, so I'm happy to fix these test cases and change the check to an asserttion as you suggested.

mattklein123 replied:
Yes please just fix the tests if they don't make sense.

I will point out that adding a header with an empty value is OK. Are you sure we don't need to support this? It seems like maybe we should support this and the original code actually has a bug?

agrawroh replied:
How about we support it for `ADD_IF_ABSENT` and `OVERWRITE_IF_EXISTS` but not for `APPEND_IF_EXISTS`? It doesn't make sense to append an empty value to the existing header.

mattklein123 replied:
I think that's fine as long as you have good test coverage for it all.

agrawroh replied:
Actually, I take what I said in 2 back. `APPEND_IF_EXISTS` also adds the header if it doesn't already exist. So I'll just drop the empty value check altogether and will add/modify tests accordingly. Thanks for your inputs!

mattklein123 reviewed (changes requested):
Thanks generally looks good. Flushing out some comments.

/wait

agrawroh commented on test/common/router/header_formatter_test.cc:
```diff
@@ -1347,6 +1344,26 @@ match: { prefix: "/new_endpoint" }
+  envoy::config::route::v3::Route route = parseRouteFromV3Yaml(yaml);
+  EXPECT_THAT_THROWS_MESSAGE(Router::HeaderParser::configure(route.request_headers_to_add()),
+                             EnvoyException,
+                             testing::HasSubstr("are set for the header key: static-header"));
```
Done!

agrawroh commented on source/common/router/header_parser.cc:
```diff
@@ -301,21 +337,34 @@ Http::HeaderTransforms HeaderParser::getHeaderTransforms(const StreamInfo::Strea
-      const std::string value = entry.formatter_->format(stream_info);
+      value = entry.formatter_->format(stream_info);
+    }
+
```
We also have a similar conversation going [here](https://github.com/envoyproxy/envoy/pull/18130#discussion_r709724490) on whether we should change these to assertions which will raise an exception if the value is empty.

agrawroh commented:
/retest

agrawroh commented:
/retest

mattklein123 commented:
@agrawroh please never force push as it makes reviews more difficult, thanks!

Also please check CI.

/wait

agrawroh commented:
> @agrawroh please never force push as it makes reviews more difficult, thanks!
> 
> Also please check CI.
> 
> /wait

@mattklein123 
Sorry about the `force-push`. I'll be careful from the next time. I resolved the CI issues. Could you please take another look whenever you get some free time?

mattklein123 commented:
@agrawroh please never force push. It makes reviews much more difficult. Thanks!

mattklein123 commented on examples/ext_authz/verify.sh:
```diff
@@ -35,6 +35,7 @@ sleep 10
 run_log "Test OPA service responds with 200"
 responds_with_header \
     "HTTP/1.1 200 OK" \
+    -X GET \
```
? This seems unrelated? Revert?

agrawroh replied:
Reverted

mattklein123 commented on source/extensions/filters/http/ext_proc/mutation_utils.cc:
```diff
@@ -62,16 +62,20 @@ void MutationUtils::applyHeaderMutations(const HeaderMutation& mutation, Http::H
-      // filter. However, the router handles this same protobuf and uses "true"
-      // as the default instead.
-      const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(sh, append, false);
+      envoy::config::core::v3::HeaderValueOption::HeaderAppendAction append_action =
```
Don't you need to support the new behavior here also? Can you move this logic into shared code so it's shared in both places with proper error handling, etc.?

agrawroh replied:
@mattklein123 I moved the code to get the append action by evaluating both the `append` as well as `append_action` flags in a utility class. Is that what you meant by "support the new behavior here"?

mattklein123 commented on test/common/router/config_impl_test.cc:
```diff
@@ -166,7 +166,9 @@ class ConfigImplTestBase {
   }
 
-  std::string responseHeadersConfig(const bool most_specific_wins, const bool append) {
+  std::string responseHeadersConfig(
```
I think you should have some DEPRECATED_FEATURE_TEST in here somewhere to make sure you test the old behavior.

agrawroh replied:
Added, Thanks!

mattklein123 commented on test/integration/header_integration_test.cc:
```diff
@@ -34,7 +34,7 @@ std::string ipSuppressEnvoyHeadersTestParamsToString(
 void disableHeaderValueOptionAppend(
```
I would add an integration test for the new behavior.

agrawroh replied:
Added, Thanks!

mattklein123 reviewed (changes requested):
Thanks, looking good! Flushing some more comments.

/wait

agrawroh commented:
Sorry, I had to force push as one if the new commits were not signed off :/

agrawroh commented:
@mattklein123 I think the scope of this PR has become huge. Will it be okay to create separate smaller PRs to do the ExtAuthZ and RateLimit changes? I can also create a separate PR for Mutation changes. It'd be easier to get reviews on smaller PRs.

mattklein123 commented:
> @mattklein123 I think the scope of this PR has become huge. Will it be okay to create separate smaller PRs to do the ExtAuthZ and RateLimit changes? I can also create a separate PR for Mutation changes. It'd be easier to get reviews on smaller PRs.

I would rather make sure it all gets done, so please just do it all and then we can split PRs if needed, thanks.

/wait

agrawroh commented:
> > @mattklein123 I think the scope of this PR has become huge. Will it be okay to create separate smaller PRs to do the ExtAuthZ and RateLimit changes? I can also create a separate PR for Mutation changes. It'd be easier to get reviews on smaller PRs.
> 
> I would rather make sure it all gets done, so please just do it all and then we can split PRs if needed, thanks.
> 
> /wait

Sounds good. I pushed the changes for RLS, ExtAuthZ, Mutations etc.

agrawroh commented:
/retest

agrawroh commented:
/retest

agrawroh commented:
@mattklein123 I think the reason why `examples/ext_authz` fails on my last commit (which I reverted) is because the proto is not synced yet.

https://github.com/envoyproxy/envoy/blob/main/examples/ext_authz/auth/grpc-service/pkg/auth/v3/auth.go#L8

This is where it fails:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc#L118

Is there a way to tackle such changes? I think I addressed all other comments.

wbpcode commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -320,12 +320,37 @@ message HeaderValueOption {
       "envoy.api.v2.core.HeaderValueOption";
 
+  // Describes the supported actions types to be used in :ref:`append_action <envoy_v3_api_field_config.core.v3.HeaderValueOption.append_action>`
+  enum HeaderAppendAction {
+    // This action will append the specified value to the existing values if the header
+    // already exists. If the header doesn't exist then this will add the header with
+    // specified key and value.
+    APPEND_IF_EXISTS = 0;
+
+    // This action will add the header if it doesn't already exist. If the header
+    // already exists then this will be a no-op.
+    ADD_IF_ABSENT = 1;
+
+    // This action will overwrite the specified value by discarding any existing values if
+    // the header already exists. If the header doesn't exist then this will add the header
+    // with specified key and value.
+    OVERWRITE_IF_EXISTS = 2;
+  }
```
The name `APPEND_IF_EXISTS` does not seem to be appropriate. It always adds header key/value to the headers although the header key is not exist.
This action behavior is not what it literally implies.
`OVERWRITE_IF_EXISTS` has a similar problem.

agrawroh replied:
Any suggestions on what could be a more appropriate name? Maybe something like "APPEND_IF_EXISTS_OR_ADD"?

wbpcode replied:
In my personal opinion, I think may be simple `APPEND` is enough. Append new header field, whether it already exists or not.
Similarly, `OVERWRITE` is enough.

wbpcode commented on envoy/http/header_map.h:
```diff
@@ -686,8 +686,15 @@ using HeaderMapPtr = std::unique_ptr<HeaderMap>;
-  std::vector<std::pair<Http::LowerCaseString, std::string>> headers_to_overwrite;
+  // List of headers which need to be added if they are absent.
+  std::vector<std::pair<Http::LowerCaseString, std::string>> headers_to_add_if_absent;
+  // List of headers for which the new value should be appended to the existing values if these
+  // headers already exist. If the headers don't exist they are added.
+  std::vector<std::pair<Http::LowerCaseString, std::string>> headers_to_append_if_exist;
+  // List of headers for which the new value should be overwritten by discarding any existing
+  // value if these headers already exist.
+  std::vector<std::pair<Http::LowerCaseString, std::string>> headers_to_overwrite_if_exist;
```
It's great to add these comments. But the renaming doesn't seem necessary.

wbpcode commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -160,18 +160,32 @@ Http::FilterHeadersStatus Filter::encode100ContinueHeaders(Http::ResponseHeaderM
     ENVOY_STREAM_LOG(
-        trace, "ext_authz filter added header(s) to the encoded response:", *encoder_callbacks_);
-    for (const auto& header : response_headers_to_add_) {
+        trace, "ext_authz filter append header(s) to the encoded response:", *encoder_callbacks_);
```
appended ?

agrawroh replied:
Changed.

wbpcode commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -160,18 +160,32 @@ Http::FilterHeadersStatus Filter::encode100ContinueHeaders(Http::ResponseHeaderM
+                   response_headers_to_add_if_absent_.size(), response_headers_to_set_.size());
+  // Headers to be added only if they are absent.
+  if (!response_headers_to_add_if_absent_.empty()) {
+    ENVOY_STREAM_LOG(trace, "ext_authz filter add header(s) if absent to the encoded response:",
```
added?

agrawroh replied:
Changed.

wbpcode commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -160,18 +160,32 @@ Http::FilterHeadersStatus Filter::encode100ContinueHeaders(Http::ResponseHeaderM
+      if (headers.get(header.first).empty()) {
+        ENVOY_STREAM_LOG(trace, "'{}':'{}'", *encoder_callbacks_, header.first.get(),
+                         header.second);
+        headers.setCopy(header.first, header.second);
```
Use `addCopy` here. `setCopy` will make an additional unnecessary `remove` which need to searching whole header map.

agrawroh replied:
Changed. Thanks!

wbpcode commented on source/extensions/filters/http/ext_proc/mutation_utils.cc:
```diff
@@ -62,19 +62,27 @@ void MutationUtils::applyHeaderMutations(const HeaderMutation& mutation, Http::H
+        switch (append_action) {
+        case envoy::config::core::v3::HeaderValueOption::ADD_IF_ABSENT:
+          if (headers.get(lcKey).empty()) {
+            headers.setCopy(lcKey, sh.header().value());
```
Use `addCopy` here for above reason.

agrawroh replied:
Changed.

wbpcode commented on source/common/router/header_parser.cc:
```diff
@@ -286,12 +288,26 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
+      // Check whether the header already exist or not. We only
+      // need to add the header if it doesn't already exist.
+      if (headers.get(key).empty()) {
         headers.setReferenceKey(key, value);
```
Update this `setReferenceKey` to `addReferenceKey` to reduce unnecessary header map searching.

agrawroh replied:
Done!

wbpcode reviewed (changes requested):
Thanks very much. It looks great overall. This work seems much more complicated than I expected. Added some comments.

mattklein123 commented:
> Is there a way to tackle such changes? I think I addressed all other comments.

What you could do is split out the API changes into a different PR, and we can merge that and then come back to this?

agrawroh commented:
@mattklein123 @wbpcode Thanks! That'd be great. Let me open a separate PR just for the API changes. Are you okay with the enum naming or want me to change anything?

https://github.com/envoyproxy/envoy/pull/18130#discussion_r714885914

mattklein123 commented:
> Are you okay with the enum naming or want me to change anything?

I would change to the more pedantic names, thanks.

agrawroh commented:
Created the PR for just the `.proto` changes: https://github.com/envoyproxy/envoy/pull/18246

agrawroh commented:
/retest

mattklein123 commented:
@agrawroh please stop force pushing. Thanks!

mattklein123 reviewed (commented):
Thanks generally this LGTM.

@esmet @dio can you please review the extauth changes?

@gbrail can you review the extproc changes?

/wait-any

gbrail commented:
LGTM for me and the ext_proc changes. Thanks!

mattklein123 commented:
> @esmet @dio can you please review the extauth changes?

@esmet @dio friendly ping on this. Please review as this is tricky code and could easily cause a CVE. Thanks!

agrawroh commented:
/retest

esmet commented on source/extensions/filters/common/ext_authz/ext_authz.h:
```diff
@@ -68,16 +68,23 @@ struct Response {
   Http::HeaderVector headers_to_append;
+  // A set of HTTP headers returned by the authorization server, that will be optionally be added
+  // to the request to the upstream server if they don't already exist.
+  Http::HeaderVector headers_to_add_if_absent;
```
Since the authorization service can see all of the original request headers (or be configured to so, in the case of an HTTP server), we shouldn't need the conditional "if absent" set since the server can determine whether it exists originally or not. If the header exists, it can just omit it from "headers_to_add" and if it doesn't exist it can be added to "headers_to_add".

For simplicity I think it would be best to drop this particular functionality. What do you think?

agrawroh replied:
Thanks, @esmet. Sure I can just remove `headers_to_add_if_absent` from here and also rename `response_headers_to_append` --> `response_headers_to_add` as before.

agrawroh replied:
@esmet What action should we take in case a request header comes with `ADD_IF_ABSENT` action? Should we throw an error saying it's not supported for request headers or just append/add it?

esmet replied:
That's a good point. I realize now that this is coming from a common API (the header value option) and the fact that using it doesn't make much sense for ext_authz (given Check requests), it still makes sense to implement it. Ignoring it or throwing an error would be odd.

I think we can just keep it, then. Sorry for the misguidance.

agrawroh replied:
No problem at all. I can revert my changes of removing it and push rest of the changes addressing your comments for a final review. Thank you so much for reviewing.

esmet commented on source/extensions/filters/common/ext_authz/ext_authz.h:
```diff
@@ -68,16 +68,23 @@ struct Response {
   // (using "addCopy") to the response sent back to the downstream client on OK auth
   // responses.
-  Http::HeaderVector response_headers_to_add;
+  Http::HeaderVector response_headers_to_append;
```
Shouldn't this stay response_headers_to_add, at least because the proto spec didn't also change? Name drift could cause confusion.

agrawroh replied:
Sure. Will revert it back to `response_headers_to_add`.

esmet commented on source/extensions/filters/common/ext_authz/ext_authz.h:
```diff
@@ -68,16 +68,23 @@ struct Response {
+  // A set of HTTP headers returned by the authorization server, will be optionally set
+  // (using "setCopy") to the response sent back to the downstream client (if they are absent)
+  // on OK auth responses.
+  Http::HeaderVector response_headers_to_add_if_absent;
```
To add color to the previous comment, this new field to conditionally remove headers is in fact appropriate since the ext_authz filter can't know ahead of time what the upstream response headers will be.

agrawroh replied:
Make sense. Will drop it as suggested. Thanks!

esmet commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -252,13 +274,14 @@ void Filter::onComplete(Filters::Common::ExtAuthz::ResponsePtr&& response) {
       // Currently it is not possible to add {{"append": "1"}, {"append": "2"}} (the intended
       // combined headers: {{"original": "true"}, {"append": "1"}, {"append": "2"}}) to the request
       // to upstream server by only sets `headers_to_append`.
```
Is this comment still true?

agrawroh replied:
That's a good question. Let me think about it. I think it can be achieved using `ADD_IF_ABSENT` now.

esmet commented on source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:
```diff
@@ -116,12 +126,22 @@ void GrpcClientImpl::toAuthzResponseHeader(
     ResponsePtr& response,
     const Protobuf::RepeatedPtrField<envoy::config::core::v3::HeaderValueOption>& headers) {
   for (const auto& header : headers) {
-    if (header.append().value()) {
```
From my understanding, `header.append().value()` will be true if and only if the append field was was present and had a true value. In this case, the old code would add to headers_to_append (i.e APPEND_IF_EXISTS_OR_ADD). In all other cases (e.g. field is missing or field is explicitly set to false) it would add to headers_to_set (i.e, OVERWRITE_IF_EXISTS_OR_ADD).

The new code gets an append_action from Http::HeaderUtility::getHeaderAppendAction(header). If the append field is not present, then the default action is returned (APPEND_IF_EXISTS_OR_ADD). I think this doesn't match the existing behavior, so I could either be missing something about how the new code works of my understanding is off - what do you think?

agrawroh replied:
Hmmm. Current behavior of when "append" is missing seems to be "true" everywhere else (Mutations etc.) which means that "APPEND_IF_EXISTS_OR_ADD" should be the right action given we want to keep things consistent.
That said, I did come across a comment assuming opposite of this and forcing "OVERWRITE" when "append" is missing (perhaps a bug?)

@esmet I'm okay either ways in case you want to change it back to "OVERWRITE" instead of "APPEND" when the "append" field is not specified.

![image](https://user-images.githubusercontent.com/12120828/135651888-1dfa1218-02c1-4e48-8067-4228d7c5f49f.jpeg)

esmet replied:
The documentation definitely does say that the default is true, but I don't think that's totally true in practice. I see three places `append` is used:

ext_authz/ext_authz_grpc_impl.cc (here) Uses the default value from protobuf (false)
```
          if (header.append().value()) {
              ...
```

ext_proc/mutation_utils.cc defaults to false with a comment noting ext_authz as precedent
```
      // Make "false" the default. This is logical and matches the ext_authz
      // filter. However, the router handles this same protobuf and uses "true"
      // as the default instead.
      const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(sh, append, false);
```

router/header_parser.cc defaults to true
```
const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(header_value_option, append, true);
```

So, I wonder if the comment was originally authored when the router was the only meaningful user of the 'append' field on HeaderValueOption. Later, when ext_authz and ext_proc adopted it, they chose the more obvious default but the comment then became stale.

If all of this makes sense (and is correct) then I think we need to update the comment for `append` to explain what's going on. Then, we need to update the ext_authz code paths to make sure there isn't a behavioral change.

agrawroh replied:
Make sense. So in the context of this PR, should we default to "OVERWRITE_IF_EXISTS_OR_ADD" here instead of "APPEND_IF_EXISTS_OR_ADD"?

esmet replied:
I believe the default needs to remain OVERWRITE_IF_EXISTS_OR_ADD. Ideally, we would have a unit test that verifies this behavior when omitting `append` before any of the changes in this PR. Then, when this PR adds a new field, the test should still pass as it does not set `append` nor `append_action`.

esmet commented on test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc:
```diff
@@ -305,21 +328,40 @@ class ExtAuthzGrpcIntegrationTest : public Grpc::GrpcClientIntegrationParamTest,
 
-      entry->mutable_append()->set_value(true);
+      entry->set_append_action(envoy::config::core::v3::HeaderValueOption::ADD_IF_ABSENT);
+      entry->clear_append();
```
IIUC append should not be set here so I think clearing it could theoretically mask issues. Your new code does validate that the append boolean and append_action fields are consistent so lets depend on that to find issues in test.

agrawroh replied:
+1 Can remove this and add another test case with none if the "append" or "append_action" set.

agrawroh commented on test/extensions/filters/common/ext_authz/test_common.h:
```diff
@@ -19,7 +19,7 @@ namespace ExtAuthz {
 struct KeyValueOption {
   std::string key;
   std::string value;
-  bool append;
```
I'll have to look closer. We might have 1-2 test cases covering that scenario but if not then I can certainly add them.

esmet commented on test/extensions/filters/common/ext_authz/ext_authz_http_impl_test.cc:
```diff
@@ -346,11 +353,16 @@ using HeaderValuePair = std::pair<const Http::LowerCaseString, const std::string
+        envoy::config::core::v3::HeaderValueOption::OVERWRITE_IF_EXISTS_OR_ADD}});
   const auto authz_response = TestCommon::makeAuthzResponse(
       CheckStatus::OK, Http::Code::OK, EMPTY_STRING, TestCommon::makeHeaderValueOption({}),
-      // By default, the value of envoy.config.core.v3.HeaderValueOption.append is true.
```
FTR I think this comment wasn't actually true (cc @dio ?) since the code was checking for `option.append().value()` without using the PROTOBUF_GET_... wrapper to default to true.

dio replied:
Yes that's true @esmet

esmet commented on test/extensions/filters/common/ext_authz/ext_authz_http_impl_test.cc:
```diff
@@ -346,11 +353,16 @@ using HeaderValuePair = std::pair<const Http::LowerCaseString, const std::string
+       {"x-downstream-ok", "1",
+        envoy::config::core::v3::HeaderValueOption::APPEND_IF_EXISTS_OR_ADD},
+       {"x-upstream-ok", "1",
+        envoy::config::core::v3::HeaderValueOption::OVERWRITE_IF_EXISTS_OR_ADD}});
```
If the test previously passed `false` for all three append values, then I think the new values should all consistently be `OVERWRITE_IF_EXISTS_OR_ADD` to make sure there aren't regressions.
